src/widget/wtrackmenu.cpp

```diff
@@ -0,0 +1,1548 @@
+        m_pImportMetadataFromMusicBrainzAct->setEnabled(oneSongSelected);
+        m_pMetadataMenu->addAction(m_pImportMetadataFromMusicBrainzAct);
+        m_pMetadataMenu->addAction(m_pExportMetadataAct);
+
+        for (const auto& updateInExternalTrackCollection : m_updateInExternalTrackCollections) {
+            ExternalTrackCollection* externalTrackCollection =
+                    updateInExternalTrackCollection.externalTrackCollection;
+            if (externalTrackCollection) {
+                updateInExternalTrackCollection.action->setEnabled(
+                        externalTrackCollection->isConnected());
+                m_pMetadataUpdateExternalCollectionsMenu->addAction(
+                        updateInExternalTrackCollection.action);
+            }
+        }
+        if (!m_pMetadataUpdateExternalCollectionsMenu->isEmpty()) {
+            m_pMetadataMenu->addMenu(m_pMetadataUpdateExternalCollectionsMenu);
+        }
+
+        for (const auto& updateInExternalTrackCollection : m_updateInExternalTrackCollections) {
+            ExternalTrackCollection* externalTrackCollection =
+                    updateInExternalTrackCollection.externalTrackCollection;
+            if (externalTrackCollection) {
+                updateInExternalTrackCollection.action->setEnabled(
+                        externalTrackCollection->isConnected());
+                m_pMetadataUpdateExternalCollectionsMenu->addAction(
+                        updateInExternalTrackCollection.action);
+            }
+        }
+        if (!m_pMetadataUpdateExternalCollectionsMenu->isEmpty()) {
+            m_pMetadataMenu->addMenu(m_pMetadataUpdateExternalCollectionsMenu);
+        }
+        // Cover art menu only applies if at least one track is selected.
+        if (indices.size()) {
+            // We load a single track to get the necessary context for the cover (we use
+            // last to be consistent with selectionChanged above).
+            QModelIndex last = indices.last();
+            CoverInfo info;
+            info.source = static_cast<CoverInfo::Source>(
+                    last.sibling(last.row(), m_iCoverSourceColumn).data().toInt());
+            info.type = static_cast<CoverInfo::Type>(
+                    last.sibling(last.row(), m_iCoverTypeColumn).data().toInt());
+            info.hash = last.sibling(last.row(), m_iCoverHashColumn).data().toUInt();
+            info.trackLocation = last.sibling(
+                    last.row(), m_iTrackLocationColumn).data().toString();
+            info.coverLocation = last.sibling(
+                    last.row(), m_iCoverLocationColumn).data().toString();
+            m_pCoverMenu->setCoverArt(info);
+            m_pMetadataMenu->addMenu(m_pCoverMenu);
+        }
+        addMenu(m_pMetadataMenu);
+    }
+
+    if (optionIsEnabled(Filter::Reset)) {
+        DEBUG_ASSERT(m_pTrackModel);
+        if (m_pTrackModel == nullptr) {
+            return;
+        }
```

Comment:
This is what our VERIFY_OR_DEBUG_ASSERT macro is for
```suggestion
        VERIFY_OR_DEBUG_ASSERT(m_pTrackModel) {
            return;
        }
```

Reply:
Don't actually commit this change. Instead work on removing WTrackMenu's dependency on TrackModel. I am just commenting here to inform you for future situations.

Reply:
Oh great. This is much more helpful.

Reply:
While most functionality can be decoupled, some functionality like Remove, Hide and Purge depend on the track model to determine if they can be shown in that menu. Also, the WTrackMenu needs the track model pointer to actually execute those operations.

Reply:
Okay. Please pass the TrackModel* to the WTrackMenu constructor and add `VERIFY_OR_DEBUG_ASSERTS` there to verify that the flags passed don't depend on having a TrackModel* if a nullptr is passed to the constructor.

Reply:
I had started off by including TrackModel* as a parameter as the constructor for WTrackMenu, but at the moment of initializing the menu in the WTrackTableView constructor, the model is not yet set (it is set by WTrackTableView::loadTrackModel slot), so TrackModel is always empty in the constructor. I am planning on moving the menu initialization in WTrackTableView to the loadTrackModel function.

Reply:
good plan

Reply:
Add
```
VERIFY_OR_DEBUG_ASSERT(m_pMenu) {
    return;
}
```
to WTrackTableView::contextMenuEvent just in case somehow there is a right click before loading the table model.